File: src/main/java/org/modelingvalue/nelumbo/patterns/OptionalPattern.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.patterns;

import java.io.Serial;
import java.util.Optional;
import java.util.function.Function;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.mutable.MutableList;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.Type;
import org.modelingvalue.nelumbo.Variable;
import org.modelingvalue.nelumbo.syntax.ParseState;
import org.modelingvalue.nelumbo.syntax.Token;
import org.modelingvalue.nelumbo.syntax.TokenType;

public class OptionalPattern extends Pattern {
    @Serial
    private static final long serialVersionUID = 3011113311569598643L;

    public OptionalPattern(Type type, List<AstElement> elements, Object... args) {
        super(type, elements, args);
    }

    protected OptionalPattern(Object[] args, OptionalPattern declaration) {
        super(args, declaration);
    }

    @Override
    protected OptionalPattern struct(Object[] array, Node declaration) {
        return new OptionalPattern(array, (OptionalPattern) declaration);
    }

    public Pattern optional() {
        return (Pattern) get(0);
    }

    @Override
    public ParseState state(ParseState next, Functor functor) {
        return optional().state(next, functor).merge(next);
    }

    @Override
    public String toString(TokenType[] previous) {
        return "<(>" + optional() + "<)?>";
    }

    @Override
    public Pattern setPresedence(List<Integer> precedence, int[] p) {
        return set(0, optional().setPresedence(precedence, p));
    }

    @Override
    public Pattern setTypes(Function<Type, Type> typeFunction) {
        return set(0, optional().setTypes(typeFunction));
    }

    @Override
    public List<Type> argTypes(List<Type> types) {
        return optional().argTypes(types);
    }

    @Override
    protected int string(List<Object> args, int ai, StringBuffer sb, TokenType[] previous, boolean alt) {
        if (args.get(ai) instanceof Optional<?> opt) {
            StringBuffer inner = new StringBuffer();
            if (opt.isPresent()) {
                int ii = optional().string(List.of(opt.get()), 0, inner, previous, false);
                if (ii < 0) {
                    return -1;
                }
            }
            sb.append(inner);
            return ai + 1;
        }
        return -1;
    }

    @Override
    protected int args(List<AstElement> elements, int i, MutableList<Object> args, boolean alt, Functor functor, Map<Variable, Type> typeArgs) {
        MutableList<Object> inner = MutableList.of(List.of());
        int ii = optional().args(elements, i, inner, false, functor, typeArgs);
        if (ii >= 0) {
            Object first = inner.toImmutable().first();
            args.add(first != null ? Optional.of(first) : Optional.empty());
            return ii;
        } else {
            args.add(Optional.empty());
            return i;
        }
    }

    @Override
    public Pattern declaration(Token token) {
        return optional().declaration(token);
    }

}

File: web-ts/src/patterns/OptionalPattern.ts
/**
 * OptionalPattern - matches a pattern zero or one time.
 * @JAVA_REF org.modelingvalue.nelumbo.patterns.OptionalPattern
 */

import { List, Map } from 'immutable';
import { TokenType } from '../syntax/TokenType';
import type { Token } from '../syntax/Token';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Variable } from '../Variable';
import { Node } from '../Node';
import { Pattern } from './Pattern';
import type { ParseState } from '../syntax/ParseState';
import type { Functor } from './Functor';

export class OptionalPattern extends Pattern {
  constructor(type: Type, elements: List<AstElement>, ...args: unknown[]) {
    super(type, elements, ...args);
  }

  protected static fromData(data: unknown[], declaration?: Node): OptionalPattern {
    const pattern = Object.create(OptionalPattern.prototype) as OptionalPattern;
    (pattern as unknown as { _data: unknown[] })._data = data;
    (pattern as unknown as { _declaration: Node })._declaration = declaration ?? pattern;
    return pattern;
  }

  protected struct(data: unknown[], declaration?: Node): OptionalPattern {
    return OptionalPattern.fromData(data, declaration ?? this.declaration());
  }

  optional(): Pattern {
    return this.get(0) as Pattern;
  }

  // @JAVA_REF OptionalPattern.state(ParseState next)
  parseState(next: ParseState): ParseState {
    return this.optional().parseState(next).merge(next);
  }

  toString(_previous?: TokenType[]): string {
    return '<(>' + this.optional() + '<)?>';
  }

  setPrecedence(precedence: number): Pattern {
    return this.set(0, this.optional().setPrecedence(precedence)) as OptionalPattern;
  }

  setTypes(typeFunction: (type: Type) => Type): Pattern {
    return this.set(0, this.optional().setTypes(typeFunction)) as OptionalPattern;
  }

  argTypes(types: List<Type>): List<Type> {
    return this.optional().argTypes(types);
  }

  string(args: List<unknown>, ai: number, sb: string[], previous: TokenType[], _alt: boolean): number {
    const opt = args.get(ai);
    // Check if it's an "Optional" (represented as undefined for empty, or a value for present)
    if (opt === undefined || opt === null) {
      // Empty optional - produce nothing
      return ai + 1;
    }

    // If wrapped in an object with isPresent/value pattern
    if (typeof opt === 'object' && 'isPresent' in opt) {
      const inner: string[] = [];
      if ((opt as { isPresent: boolean }).isPresent) {
        const val = (opt as { isPresent: boolean; value: unknown }).value;
        const ii = this.optional().string(List([val]), 0, inner, previous, false);
        if (ii < 0) {
          return -1;
        }
      }
      sb.push(...inner);
      return ai + 1;
    }

    // Try to match the value directly
    const inner: string[] = [];
    const ii = this.optional().string(List([opt]), 0, inner, previous, false);
    if (ii >= 0) {
      sb.push(...inner);
    }
    return ai + 1;
  }

  extractArgs(
    elements: List<AstElement>,
    i: number,
    args: unknown[],
    _alt: boolean,
    functor: Functor,
    typeArgs: Map<Variable, Type>
  ): number {
    const inner: unknown[] = [];
    const ii = this.optional().extractArgs(elements, i, inner, true, functor, typeArgs);
    if (ii >= 0) {
      const first = inner[0];
      args.push(first !== undefined ? Optional.of(first) : Optional.empty());
      return ii;
    } else {
      args.push(Optional.empty());
      return i;
    }
  }

  tokenDeclaration(token: Token): Pattern | null {
    return this.optional().tokenDeclaration(token);
  }
}

/**
 * Optional wrapper type (TypeScript equivalent of Java Optional).
 */
export class Optional<T> {
  private readonly _value: T | undefined;
  private readonly _present: boolean;

  private constructor(value: T | undefined, present: boolean) {
    this._value = value;
    this._present = present;
  }

  static of<T>(value: T): Optional<T> {
    return new Optional(value, true);
  }

  static empty<T>(): Optional<T> {
    return new Optional<T>(undefined, false);
  }

  isPresent(): boolean {
    return this._present;
  }

  isEmpty(): boolean {
    return !this._present;
  }

  get(): T | undefined {
    return this._value;
  }

  orElse(defaultValue: T): T {
    return this._present ? this._value! : defaultValue;
  }
}

export function some<T>(value: T): Optional<T> {
  return Optional.of(value);
}

export function none<T>(): Optional<T> {
  return Optional.empty<T>();
}
